hwloc/topology-rsmi.c

```diff
@@ -0,0 +1,243 @@
+
+  if (rsmi_rc != RSMI_STATUS_SUCCESS) {
+    return -1;
+  }
+  lanespeed_raw = bandwidth.transfer_rate.frequency[bandwidth.transfer_rate.current]; // T/s
+  lanespeed = lanespeed_raw <= 5000000000 ? (lanespeed_raw * 8)/10 : (lanespeed_raw * 128)/130; // bits/s
+  lanes = bandwidth.lanes[bandwidth.transfer_rate.current];
+  *linkspeed = (lanespeed * lanes) / 8000000000; // GB/s
+  return 0;
+}
+
+/*
+ * Get the Unique ID of the GPU
+ *
+ * dv_ind  (IN) The device index
+ * buffer  (OUT) Unique ID of GPU devices
+ */
+static int get_device_unique_id(uint32_t dv_ind, char *buffer)
+{
+  uint64_t id;
+  rsmi_status_t rsmi_rc = rsmi_dev_unique_id_get(dv_ind, &id);
+
+  if (rsmi_rc != RSMI_STATUS_SUCCESS) {
+    return -1;
+  }
+  sprintf(buffer, "%lx", id);
+  return 0;
+}
+
+/*
+ * Get the serial number of the GPU
+ *
+ * dv_ind  (IN) The device index
+ * serial  (OUT) Serial number of GPU devices
+ * size    (IN) Length of the caller provided buffer
+ */
+static int get_device_serial_number(uint32_t dv_ind, char *serial, unsigned int size)
+{
+  rsmi_status_t rsmi_rc = rsmi_dev_serial_number_get(dv_ind, serial, size);
+
+  if (rsmi_rc != RSMI_STATUS_SUCCESS) {
+    return -1;
+  }
+  return 0;
+}
+
+static int
+hwloc_rsmi_discover(struct hwloc_backend *backend, struct hwloc_disc_status *dstatus)
+{
+  /*
+   * This backend uses the underlying OS.
+   * However we don't enforce topology->is_thissystem so that
+   * we may still force use this backend when debugging with !thissystem.
+   */
+
+  struct hwloc_topology *topology = backend->topology;
+  enum hwloc_type_filter_e filter;
+  rsmi_status_t ret;
+  unsigned nb, i;
+
+  assert(dstatus->phase == HWLOC_DISC_PHASE_IO);
+
+  hwloc_topology_get_type_filter(topology, HWLOC_OBJ_OS_DEVICE, &filter);
+  if (filter == HWLOC_TYPE_FILTER_KEEP_NONE)
+    return 0;
+
+  rsmi_init(0);
+
+  ret = rsmi_num_monitor_devices(&nb);
+  if (RSMI_STATUS_SUCCESS != ret || !nb) {
+    if (RSMI_STATUS_SUCCESS != ret)
+      fprintf(stderr, "RSMI: Failed to get device count\n");
+    rsmi_shut_down();
+    return 0;
+  }
+
+  for (i=0; i<nb; i++) {
+    uint64_t bdfid = 0;
+    unsigned domain, device, bus, func;
+    hwloc_obj_t osdev, parent;
+    char buffer[64];
+
+    osdev = hwloc_alloc_setup_object(topology, HWLOC_OBJ_OS_DEVICE, HWLOC_UNKNOWN_INDEX);
+    snprintf(buffer, sizeof(buffer), "rsmi%u", i);
+    osdev->name = strdup(buffer);
+    osdev->depth = HWLOC_TYPE_DEPTH_UNKNOWN;
+    osdev->attr->osdev.type = HWLOC_OBJ_OSDEV_GPU;
+
+    hwloc_obj_add_info(osdev, "Backend", "RSMI");
+    hwloc_obj_add_info(osdev, "GPUVendor", "AMD");
+
+    buffer[0] = '\0';
+    if (get_device_name(i, buffer, sizeof(buffer)) == -1)
+      buffer[0] = '\0';
+    hwloc_obj_add_info(osdev, "GPUModel", buffer);
+
+    buffer[0] = '\0';
+    if (get_device_serial_number(i, buffer, sizeof(buffer)) == 0)
+      hwloc_obj_add_info(osdev, "AMDSerial", buffer);
+
```

Comment:
if buffer can be left empty, add "&& buffer[0]" at the end of the if() above

Reply:
buffer can contain some data, even so call failed.

Reply:
What I am getting on Fiji is not a failure but an empty buffer, so I wanted to ignore this empty AMDSerial.

Reply:
I didn't that. Ok, I will add check.

Thanks

Mike